### PR TITLE
Adding support for ApiNamespace 

### PR DESCRIPTION
**Pull request by dgaeta**

### CODE REVIEW

File: api-extractor/src/definitions/ApiField.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import * as ts from 'typescript';
+import { ApiItemKind, IApiItemOptions } from './ApiItem';
+import ApiMember from './ApiMember';
+
+/**
+ * This class is part of the ApiItem abstract syntax tree. It represents fields of
+ * an ApiNamespace.
+ */
+class ApiField extends ApiMember {
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiField.ts]: >ApiField [](start = 6, length = 8)

Test comment

File: api-extractor/src/definitions/ApiNamespace.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        const followedSymbol: ts.Symbol = this.followAliases(exportSymbol);
+
+        if (!followedSymbol.declarations) {
+          this.reportWarning(`Definition with no declarations: ${exportSymbol.name}`);
+          continue;
+        }
+
+        if (!(followedSymbol.flags === ts.SymbolFlags.BlockScopedVariable)) {
+          this.reportWarning(`Unsupported export "${exportSymbol.name}" ` +
+            'ApiNamespace only supports properties.');
```
Comment (dgaeta) [inline api-extractor/src/definitions/ApiNamespace.ts]: @pgonzal

File: api-extractor/src/definitions/ApiNamespace.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+        // Since we are imposing that the items within a namespace be
+        // const properties we are only taking the first declaration.
+        // If we decide to add support for other types within a namespace
+        // we will have for evaluate each declaration.
+        const declaration: ts.Declaration = followedSymbol.getDeclarations()[0];
+
+        if (declaration.parent.flags !== ts.NodeFlags.Const) {
+          this.reportWarning(`Export "${exportSymbol.name}" must possess the "const" modifier`);
+          continue;
```
Comment (dgaeta) [inline api-extractor/src/definitions/ApiNamespace.ts]: @pgonzal can you review this warning message

File: api-extractor/src/definitions/ApiNamespace.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        const propertySignature: ts.PropertySignature = declaration as ts.PropertySignature;
+        const type: string = propertySignature.type.getText();
+        if (allowedTypes.indexOf(type) < 0) {
+          this.reportWarning(`Export "${exportSymbol.name}" must of type "string", "number" or "boolean"`);
+          continue;
```
Comment (dgaeta) [inline api-extractor/src/definitions/ApiNamespace.ts]: @pgonzal can you review this warning message

File: api-extractor/src/definitions/ApiField.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  public value: string;
+
+  constructor(options: IApiItemOptions) {
+    super(options);
+    this.kind = ApiItemKind.Field;
+
+    const propertySignature: ts.PropertySignature = options.declaration as ts.PropertySignature;
+    this.type = propertySignature.type.getText();
+    this.name = propertySignature.name.getText();
+    this.value = propertySignature.initializer.getText(); // value of the export
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiField.ts]: >ropertySignature.initializer.getText(); // value of the export [](start = 18, length = 62)

What happens if there isn't an initializer?

File: api-extractor/src/definitions/ApiField.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  public type: string;
+  public name: string;
+  public value: string;
+
+  constructor(options: IApiItemOptions) {
+    super(options);
+    this.kind = ApiItemKind.Field;
+
+    const propertySignature: ts.PropertySignature = options.declaration as ts.PropertySignature;
+    this.type = propertySignature.type.getText();
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiField.ts]: >type [](start = 34, length = 4)

What if they don't declare the type?  Might be worth testing that case

File: api-extractor/src/definitions/ApiItem.ts [hunk 7]
```diff
@@ -59,7 +59,15 @@ export enum ApiItemKind {
...
    * A Typescript class constructor function.
    */
-  Constructor = 10
+  Constructor = 10,
+  /**
+   * A Typescript namespace.
+   */
+  Namespace = 11,
+  /**
+   * An abstract type field.
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiItem.ts]: >An abstract type field. [](start = 5, length = 23)

I wouldn't call this "abstract".  In C# a "field" is a language element that stores a simple data value (without any getter/setter function).

Seems like the TypeScript compiler API calls this a "BlockScopedVariable".

You could consider renaming ApiField -> ApiModuleVariable

File: api-extractor/src/definitions/ApiNamespace.ts [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+import ApiItem, { ApiItemKind, IApiItemOptions } from './ApiItem';
+import ApiItemContainer from './ApiItemContainer';
+import { IExportedSymbol } from '../IExportedSymbol';
+
+const allowedTypes: string[] = ['string', 'number', 'boolean'];
+
+/**
+  * This class is part of the ApiItem abstract syntax tree. It represents exports of
+  * a namespace, the exports can be classes, interfaces, type literals, enums, functions,
+  * and properties.
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiNamespace.ts]: This isn't true any more

File: api-extractor/src/definitions/ApiNamespace.ts (hunk 2 above)
Comment (pgonzal) [inline api-extractor/src/definitions/ApiNamespace.ts]: "Unsupported export ____. Currently the "namespace" block only supports constant variables."

---
In reply to: [114371996](https://github.com/Microsoft/web-build-tools/pull/167#discussion_r114371996) [](ancestors = 114371996)

File: api-extractor/src/definitions/ApiNamespace.ts (hunk 3 above)
Comment (pgonzal) [inline api-extractor/src/definitions/ApiNamespace.ts]: "Export ___ is missing the "const" modifier.  Currently the "namespace" block only supports constant variables."

---
In reply to: [114372232](https://github.com/Microsoft/web-build-tools/pull/167#discussion_r114372232) [](ancestors = 114372232)

File: api-extractor/src/definitions/ApiNamespace.ts [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+   *
+   * For example, suppose a class is defined as "export default class MyClass { }"
+   * but exported from the package's index.ts like this:
+   *
+   *    export { default as _MyClass } from './MyClass';
+   *
+   * In this example, given the symbol for _MyClass, getExportedSymbolName() will return
+   * the string "MyClass".
+   */
+  public tryGetExportedSymbolName(symbol: ts.Symbol): string {
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiNamespace.ts]: >tryGetExportedSymbolName [](start = 9, length = 24)

Why do we need this?  I would expect ApiNamespace to be more like ApiStructuredType than ApiPackage.

File: api-extractor/src/definitions/ApiNamespace.ts [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    }
+    return undefined;
+  }
+
+  /**
+   * Find a member in this namespace by name and return it if found.
+   *
+   * @param memberName - the name of the exported ApiItem
+   */
+  public getMemberItem(memberName: string): ApiItem {
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiNamespace.ts]: >getMemberItem [](start = 9, length = 13)

Should this be a member of ApiItemContainer?

File: api-extractor/src/definitions/ApiNamespace.ts [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    if (this.memberItems.has(memberName)) {
+      return this.memberItems.get(memberName);
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiNamespace.ts]: >get [](start = 30, length = 3)

Map.get() already returns undefined if the item was not found in the collection

**Comment by pgonzal** [comment]
:clock1:

File: api-extractor/src/definitions/ApiField.ts (hunk 5 above)
Comment (dgaeta) [inline api-extractor/src/definitions/ApiField.ts]: This isn't allowed in typescript, "[ts] 'const' declarations must be initialized". If the variable declaration is not 'const', then our system will report a warning, at which time the the compiler will require the const to be initialized with a value.

File: api-extractor/testInputs/example4/src/NamespaceExport.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
+import MyClass from './MyClass';
+
+/**
+ * This is a test for the namespace API type. It should be supported 
+ * and appear in the *api.ts file. At the moment we only support 
+ * a table of primitive types.
+ * 
+ * @public
+ */
+export namespace NamespaceExport {
```
Comment (pgonzal) [inline api-extractor/testInputs/example4/src/NamespaceExport.ts]: >namespace [](start = 7, length = 9)

You could also add a real-world test this by converting RushConstants to a namespace:

[https://github.com/Microsoft/web-build-tools/blob/master/rush/rush-lib/src/RushConstants.ts](https://github.com/Microsoft/web-build-tools/blob/master/rush/rush-lib/src/RushConstants.ts)

File: api-extractor/src/definitions/ApiNamespace.ts [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+import ApiModuleVariable from './ApiModuleVariable';
+import ApiItem, { ApiItemKind, IApiItemOptions } from './ApiItem';
+import ApiItemContainer from './ApiItemContainer';
+import { IExportedSymbol } from '../IExportedSymbol';
+
+const allowedTypes: string[] = ['string', 'number', 'boolean'];
+
+/**
+  * This class is part of the ApiItem abstract syntax tree. It represents exports of
+  * a namespace, the exports can be module variable constants of type "string", "boolean" or "number.
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiNamespace.ts]: >. [](start = 100, length = 1)

missing quote

File: api-extractor/src/definitions/ApiNamespace.ts [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+  * An ApiNamespace is defined using TypeScript's "namespace" keyword.
+  *
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiNamespace.ts]: > * [](start = 1, length = 2)

@remarks

File: api-extractor/src/definitions/ApiModuleVariable.ts [hunk 15]
```diff
@@ -0,0 +1,25 @@
+import * as ts from 'typescript';
+import { ApiItemKind, IApiItemOptions } from './ApiItem';
+import ApiMember from './ApiMember';
+
+/**
+ * This class is part of the ApiItem abstract syntax tree. It represents fields of
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiModuleVariable.ts]: >fields [](start = 73, length = 6)

"It represents variables that are exported by an ApiNamespace (or potentially in the future an ApiPackage).  The variables have a name, a type, and an initializer.  The ApiNamespace implementation currently requires them to use a primitive type and be declared as "const"."

File: api-extractor/testInputs/example4/src/NamespaceExport.ts (hunk 12 above)
Comment (dgaeta) [inline api-extractor/testInputs/example4/src/NamespaceExport.ts]: rush-constants-output

File: api-extractor/src/definitions/ApiNamespace.ts [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+          continue;
+        }
+
+        // Typescript VariableDeclaration exist within a VariableDeclarationList,
+        // the VariableDeclarationList exists within a VariableStatement and
+        // this is where the JsDoc comment Node exists.
+        // If there is no parent or grandparent of this VariableDeclartion then
+        // we do not know how to obtain the JsDoc comment.
+        let jsDocNode: ts.Node;
+        if (!declaration.parent || !declaration.parent.parent) {
```
Comment (pgonzal) [inline api-extractor/src/definitions/ApiNamespace.ts]: >) { [](start = 61, length = 3)

|| !declaration.parent.parent.kind == VariableStatement

**Comment by pgonzal** [comment]
:clock1:

**Comment by pgonzal** [comment]
:shipit:

**Pull request merged**